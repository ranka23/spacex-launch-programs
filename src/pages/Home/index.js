/* eslint-disable @typescript-eslint/explicit-module-boundary-types */

/*
 * Use .js extension for code-splitting file
 */

import React from "react";
import loadable from "@loadable/component";

import { Loading, ErrorBoundary } from "../../components";

const Home = loadable(() => import("./Home"), {
  fallback: <Loading />,
});

export default (props) => (
  <ErrorBoundary>
    <Home {...props} />
  </ErrorBoundary>
);
