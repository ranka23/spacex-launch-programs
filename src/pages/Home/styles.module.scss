.home {
  padding: 0 15px;
  max-width: 1440px;
  margin: 0 auto;

}
@media only screen and (min-width: 700px) {
  .content_container {
    display: flex;
  }
}

.sidebar_container {
  position: sticky;
}

.content_container {
  margin: auto;
  width: 100%;
}

@media only screen and (max-width: 850px) {
  .list_items_container {
    display: grid;
    grid-template-columns: repeat(1, 1fr);
  }
}

@media only screen and (min-width: 850px) and (max-width: 1180px) {
  .list_items_container {
    display: grid;
    grid-template-columns: repeat(2, 1fr);

  }
}

@media only screen and (min-width: 1180px) and (max-width: 1440px) {
  .list_items_container {
    display: grid;
    grid-template-columns: repeat(3, 1fr);
  }
}

@media only screen and (min-width: 1440px) {
  .list_items_container {
    display: grid;
    grid-template-columns: repeat(4, 1fr);
  }
}

.list_items {
  margin: 10px;
}
