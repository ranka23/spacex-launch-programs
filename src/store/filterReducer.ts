import { fetchFilteredList } from "./spaceXData";
import { AppThunk } from ".";

interface FilterState {
  yearValue: string;
  launchValue: boolean | null;
  landingValue: boolean | null;
}

interface yearClickedAction {
  type: "YEAR_CLICKED";
  value: string;
}

interface launchClickedAction {
  type: "LAUNCH_CLICKED";
  value: boolean;
}

interface landingClickedAction {
  type: "LANDING_CLICKED";
  value: boolean;
}

type FilterActions =
  | yearClickedAction
  | launchClickedAction
  | landingClickedAction;

const initialState: FilterState = {
  yearValue: "",
  landingValue: null,
  launchValue: null,
};

export default (state = initialState, action: FilterActions): FilterState => {
  switch (action.type) {
    case "YEAR_CLICKED":
      return {
        ...state,
        yearValue: action.value,
      };
    case "LAUNCH_CLICKED":
      return {
        ...state,
        launchValue: action.value,
      };
    case "LANDING_CLICKED":
      return {
        ...state,
        landingValue: action.value,
      };
    default:
      return state;
  }
};

/*  
    These actions ascertain the values clicked in the Filters Sidebar
    and also make calls to api as per the values received.
*/
export const setYearValue = (value: string): AppThunk => {
  return async (dispatch) => {
    dispatch({ type: "YEAR_CLICKED", value });
    dispatch(fetchFilteredList());
  };
};

export const setLaunchValue = (value: boolean): AppThunk => {
  return async (dispatch) => {
    dispatch({ type: "LAUNCH_CLICKED", value });
    dispatch(fetchFilteredList());
  };
};

export const setLandingValue = (value: boolean): AppThunk => {
  return async (dispatch) => {
    dispatch({ type: "LANDING_CLICKED", value });
    dispatch(fetchFilteredList());
  };
};
