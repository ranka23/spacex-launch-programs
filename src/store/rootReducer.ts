import { connectRouter } from "connected-react-router";
import { History } from "history";
import filterButton from "./filterReducer";
import spaceXList from "./spaceXData";

// Use inferred return type for making correctly Redux types
// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export default (history: History) => ({
  spaceXList,
  filterButton,
  router: connectRouter(history) as any,
});
