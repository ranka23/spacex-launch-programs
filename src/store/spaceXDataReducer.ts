import { PayloadAction, createSlice } from "@reduxjs/toolkit";

import { AppThunk, AppState } from ".";
import { LaunchData, getLaunchList, getFilterList } from "../services/api";
import getUrl from "../utils/getUrl";

interface SpaceXData {
  readyStatus: string;
  items: LaunchData[];
  error: string | null;
}

export const initialState: SpaceXData = {
  readyStatus: "invalid",
  items: [],
  error: null,
};

const spaceXList = createSlice({
  name: "launchDataList",
  initialState,
  reducers: {
    getRequest: (state: SpaceXData) => {
      state.readyStatus = "request";
    },
    getSuccess: (state, { payload }: PayloadAction<LaunchData[]>) => {
      state.readyStatus = "success";
      state.items = payload;
    },
    getFailure: (state, { payload }: PayloadAction<string>) => {
      state.readyStatus = "failure";
      state.error = payload;
    },
    getFilteredList: (state, { payload }: PayloadAction<LaunchData[]>) => {
      state.readyStatus = "request";
      state.items = payload;
    },
  },
});

export default spaceXList.reducer;
export const {
  getRequest,
  getSuccess,
  getFailure,
  getFilteredList,
} = spaceXList.actions;

export const fetchSpaceXList = (): AppThunk => async (dispatch, getState) => {
  const { launchValue } = getState().filterButton;
  const { landingValue } = getState().filterButton;
  const { yearValue } = getState().filterButton;

  const url = getUrl(yearValue, launchValue, landingValue);


  dispatch(getRequest());

  const { error, data } = await getLaunchList(url);

  if (error) {
    dispatch(getFailure(error.message));
  } else {
    dispatch(getSuccess(data as LaunchData[]));
  }
};

const shouldFetchLaunchList = (state: AppState) =>
  state.spaceXList.readyStatus !== "success";

export const fetchSpaceXListIfNeed = (): AppThunk => (dispatch, getState) => {
  if (shouldFetchLaunchList(getState())) return dispatch(fetchSpaceXList());

  return null;
};

export const fetchFilteredList = (): AppThunk => async (dispatch, getState) => {
  const { launchValue } = getState().filterButton;
  const { landingValue } = getState().filterButton;
  const { yearValue } = getState().filterButton;

  const url = getUrl(yearValue, launchValue, landingValue);
  dispatch(getRequest());

  const { error, data } = await getFilterList(url);

  if (error) {
    dispatch(getFailure(error.message));
  } else {
    dispatch(getSuccess(data as LaunchData[]));
  }
};
