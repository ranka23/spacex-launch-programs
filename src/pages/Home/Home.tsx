import React, { memo, useEffect } from "react";
import { Helmet } from "react-helmet";
import { shallowEqual, useDispatch, useSelector } from "react-redux";
import Item from "../../components/Items";
import Sidebar from "../../components/Sidebar";
import { AppState } from "../../store";
import { fetchSpaceXListIfNeed } from "../../store/spaceXData";
import styles from "./styles.module.scss";

const Home = (): JSX.Element => {
  const dispatch = useDispatch();
  const { readyStatus, items } = useSelector(
    ({ spaceXList }: AppState) => spaceXList,
    shallowEqual
  );

  useEffect(() => {
    dispatch(fetchSpaceXListIfNeed());
  }, [dispatch]);

  const renderList = () => {
    if (!readyStatus || readyStatus === "invalid" || readyStatus === "request")
      return <p>Loading...</p>;

    if (readyStatus === "failure") return <p>Failed to load SpaceX data.</p>;

    return items.map((item) => {
      let landing;

      if (item.launch_success === false) {
        landing = false;
      }

      landing = item.rocket.first_stage.cores[0].land_success;

      if (landing === null) {
        landing = false;
      }

      return (
        <div className={styles.list_items} key={item.flight_number}>
          <Item
            missionImage={item.links.mission_patch_small}
            missionName={item.mission_name}
            flightNumber={item.flight_number}
            missionId={item.mission_id}
            launchYear={item.launch_year}
            successfulLaunch={item.launch_success}
            successfulLanding={landing}
          />
        </div>
      );
    });
  };

  return (
    <div className={styles.home}>
      <Helmet title="Home" />
      <div className={styles.content_container}>
        <Sidebar />
        <div className={styles.list_items_container}>{renderList()}</div>
      </div>
    </div>
  );
};

export default memo(Home);
