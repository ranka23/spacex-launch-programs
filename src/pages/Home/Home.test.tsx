import React from "react";
import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";

import { fetchSpaceXListIfNeed } from "../../store/spaceXData";

import mockStore from "../../utils/mockStore";
import Home from "./Home";

describe("<Home />", () => {
  const renderHelper = (reducer = { readyStatus: "invalid" }) => {
    const { dispatch, ProviderWithStore } = mockStore({
      spaceXList: reducer,
      filterButton: {
        launchValue: false,
        landingValue: true,
      },
    });
    const { container } = render(
      <ProviderWithStore>
        <MemoryRouter>
          <Home />
        </MemoryRouter>
      </ProviderWithStore>
    );

    return { dispatch, firstChild: container.firstChild };
  };

  it("Should fetch data when page loads", () => {
    const { dispatch } = renderHelper();

    expect(dispatch).toHaveBeenCalledTimes(1);
    expect(dispatch.mock.calls[0][0].toString()).toBe(
      fetchSpaceXListIfNeed().toString()
    );
  });

  it("Renders the loading status if data invalid", () => {
    expect(renderHelper().firstChild).toMatchSnapshot();
  });

  it("Renders the loading status when requesting data", () => {
    const reducer = { readyStatus: "request" };

    expect(renderHelper(reducer).firstChild).toMatchSnapshot();
  });

  it("Renders the <List /> if loading was successful", () => {
    const reducer = {
      readyStatus: "success",
      items: [
        {
          flight_number: 1,
          mission_name: "FalconSat",
          mission_id: ["232", "234234"],
          launch_year: "2006",
          rocket: {
            first_stage: {
              cores: [
                {
                  land_success: false,
                },
              ],
            },
          },
          launch_success: false,
          links: {
            mission_patch: "https://images2.imgbox.com/40/e3/GypSkayF_o.png",
          },
        },
      ],
    };

    expect(renderHelper(reducer).firstChild).toMatchSnapshot();
  });
});
