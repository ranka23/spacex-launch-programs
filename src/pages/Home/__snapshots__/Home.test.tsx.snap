// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> Renders the <List /> if loading was successful 1`] = `
<div
  class="home"
>
  <div
    class="content_container"
  >
    <div
      class="sidebar_container"
    >
      <h4
        class="filter_title"
      >
        Filters
      </h4>
      <div
        class="filter_container"
      >
        <h5>
          Launch Year
        </h5>
        <div
          class="toggle_button_container"
        >
          <button
            class="button null"
            type="button"
          >
            2006
          </button>
          <button
            class="button null"
            type="button"
          >
            2007
          </button>
          <button
            class="button null"
            type="button"
          >
            2008
          </button>
          <button
            class="button null"
            type="button"
          >
            2009
          </button>
          <button
            class="button null"
            type="button"
          >
            2010
          </button>
          <button
            class="button null"
            type="button"
          >
            2011
          </button>
          <button
            class="button null"
            type="button"
          >
            2012
          </button>
          <button
            class="button null"
            type="button"
          >
            2013
          </button>
          <button
            class="button null"
            type="button"
          >
            2014
          </button>
          <button
            class="button null"
            type="button"
          >
            2015
          </button>
          <button
            class="button null"
            type="button"
          >
            2016
          </button>
          <button
            class="button null"
            type="button"
          >
            2017
          </button>
          <button
            class="button null"
            type="button"
          >
            2018
          </button>
          <button
            class="button null"
            type="button"
          >
            2019
          </button>
          <button
            class="button null"
            type="button"
          >
            2020
          </button>
        </div>
      </div>
      <div
        class="filter_container"
      >
        <h5>
          Successful Launch
        </h5>
        <div
          class="toggle_button_container"
        >
          <button
            class="button null"
            type="button"
          >
            true
          </button>
          <button
            class="button button_active"
            type="button"
          >
            false
          </button>
        </div>
      </div>
      <h5>
        Successful Landing
      </h5>
      <div
        class="toggle_button_container"
      >
        <button
          class="button button_active"
          type="button"
        >
          true
        </button>
        <button
          class="button null"
          type="button"
        >
          false
        </button>
      </div>
    </div>
    <div
      class="list_items_container"
    >
      <div
        class="list_items"
      >
        <div
          class="card_container"
        >
          <img
            alt="spaceX Mission Badge"
            class="item_image"
            src="https://images2.imgbox.com/40/e3/GypSkayF_o.png"
          />
          <h4
            class="item_name"
          >
            FalconSat
             #
            1
          </h4>
          <div
            class="Flex value_container"
            data-test="flex"
          >
            <p
              class="value_title"
            >
              Mission id
              :
            </p>
            <ul
              class="list_value"
            >
              <li
                class="value"
              >
                232
              </li>
              <li
                class="value"
              >
                234234
              </li>
            </ul>
          </div>
          <div
            class="Flex value_container"
            data-test="flex"
          >
            <p
              class="value_title"
            >
              Launch Year
              :
            </p>
            <span
              class="value"
            >
              2006
            </span>
          </div>
          <div
            class="Flex value_container"
            data-test="flex"
          >
            <p
              class="value_title"
            >
              Successful Launch
              :
            </p>
            <span
              class="value"
            >
              false
            </span>
          </div>
          <div
            class="Flex value_container"
            data-test="flex"
          >
            <p
              class="value_title"
            >
              Successful Landing
              :
            </p>
            <span
              class="value"
            >
              false
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<Home /> Renders the loading status if data invalid 1`] = `
<div
  class="home"
>
  <div
    class="content_container"
  >
    <div
      class="sidebar_container"
    >
      <h4
        class="filter_title"
      >
        Filters
      </h4>
      <div
        class="filter_container"
      >
        <h5>
          Launch Year
        </h5>
        <div
          class="toggle_button_container"
        >
          <button
            class="button null"
            type="button"
          >
            2006
          </button>
          <button
            class="button null"
            type="button"
          >
            2007
          </button>
          <button
            class="button null"
            type="button"
          >
            2008
          </button>
          <button
            class="button null"
            type="button"
          >
            2009
          </button>
          <button
            class="button null"
            type="button"
          >
            2010
          </button>
          <button
            class="button null"
            type="button"
          >
            2011
          </button>
          <button
            class="button null"
            type="button"
          >
            2012
          </button>
          <button
            class="button null"
            type="button"
          >
            2013
          </button>
          <button
            class="button null"
            type="button"
          >
            2014
          </button>
          <button
            class="button null"
            type="button"
          >
            2015
          </button>
          <button
            class="button null"
            type="button"
          >
            2016
          </button>
          <button
            class="button null"
            type="button"
          >
            2017
          </button>
          <button
            class="button null"
            type="button"
          >
            2018
          </button>
          <button
            class="button null"
            type="button"
          >
            2019
          </button>
          <button
            class="button null"
            type="button"
          >
            2020
          </button>
        </div>
      </div>
      <div
        class="filter_container"
      >
        <h5>
          Successful Launch
        </h5>
        <div
          class="toggle_button_container"
        >
          <button
            class="button null"
            type="button"
          >
            true
          </button>
          <button
            class="button button_active"
            type="button"
          >
            false
          </button>
        </div>
      </div>
      <h5>
        Successful Landing
      </h5>
      <div
        class="toggle_button_container"
      >
        <button
          class="button button_active"
          type="button"
        >
          true
        </button>
        <button
          class="button null"
          type="button"
        >
          false
        </button>
      </div>
    </div>
    <div
      class="list_items_container"
    >
      <p>
        Loading...
      </p>
    </div>
  </div>
</div>
`;

exports[`<Home /> Renders the loading status when requesting data 1`] = `
<div
  class="home"
>
  <div
    class="content_container"
  >
    <div
      class="sidebar_container"
    >
      <h4
        class="filter_title"
      >
        Filters
      </h4>
      <div
        class="filter_container"
      >
        <h5>
          Launch Year
        </h5>
        <div
          class="toggle_button_container"
        >
          <button
            class="button null"
            type="button"
          >
            2006
          </button>
          <button
            class="button null"
            type="button"
          >
            2007
          </button>
          <button
            class="button null"
            type="button"
          >
            2008
          </button>
          <button
            class="button null"
            type="button"
          >
            2009
          </button>
          <button
            class="button null"
            type="button"
          >
            2010
          </button>
          <button
            class="button null"
            type="button"
          >
            2011
          </button>
          <button
            class="button null"
            type="button"
          >
            2012
          </button>
          <button
            class="button null"
            type="button"
          >
            2013
          </button>
          <button
            class="button null"
            type="button"
          >
            2014
          </button>
          <button
            class="button null"
            type="button"
          >
            2015
          </button>
          <button
            class="button null"
            type="button"
          >
            2016
          </button>
          <button
            class="button null"
            type="button"
          >
            2017
          </button>
          <button
            class="button null"
            type="button"
          >
            2018
          </button>
          <button
            class="button null"
            type="button"
          >
            2019
          </button>
          <button
            class="button null"
            type="button"
          >
            2020
          </button>
        </div>
      </div>
      <div
        class="filter_container"
      >
        <h5>
          Successful Launch
        </h5>
        <div
          class="toggle_button_container"
        >
          <button
            class="button null"
            type="button"
          >
            true
          </button>
          <button
            class="button button_active"
            type="button"
          >
            false
          </button>
        </div>
      </div>
      <h5>
        Successful Landing
      </h5>
      <div
        class="toggle_button_container"
      >
        <button
          class="button button_active"
          type="button"
        >
          true
        </button>
        <button
          class="button null"
          type="button"
        >
          false
        </button>
      </div>
    </div>
    <div
      class="list_items_container"
    >
      <p>
        Loading...
      </p>
    </div>
  </div>
</div>
`;
